File: first-architecture/Invokers/dependency-providers/InvokerACL/function/ACL-js-function/accessControl.js
// Define the Access Control List (ACL)
const acl = {
    admin: ["read", "write", "delete"],
    editor: ["read", "write"],
    viewer: ["read"]
};

const roles = {
    Bob: "admin",
    Alice: "editor",
    John: "viewer"
}

/**
 * Function to check if a user role has a specific permission
 * @param {string} user - The user name
 * @param {string} permission - The permission to check (e.g., "read", "write", "delete")
 * @returns {boolean} - True if the role has the permission, false otherwise
 */
function hasAccess(user, permission) {
    //Get the role
    const role = roles[user];

    // Get the permissions for the role
    const permissions = acl[role];

    // If the role is not found in the ACL, deny access
    if (!permissions) {
        return false;
    }

    // Check if the permission exists in the role's permissions
    return permissions.includes(permission);
}

// Get input from the command line
let params
try {
    params = JSON.parse(process.argv[2])
}catch (error) {
    console.error("Error parsing input")
    process.exit(1)
}


// Check access and display the result
const accessGranted = hasAccess(params.user, params.permission);
console.log(`${accessGranted}`);

File: first-architecture/Invokers/dependency-providers/InvokerDP/go/ACL-go-function/accessControl.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

// Define the Access Control List (ACL)
var acl = map[string][]string{
	"admin":  {"read", "write", "delete"},
	"editor": {"read", "write"},
	"viewer": {"read"},
}

var roles = map[string]string{
	"Bob":   "admin",
	"Alice": "editor",
	"John":  "viewer",
}

// Function to check if a user role has a specific permission
func hasAccess(user string, permission string) bool {
	// Get the role
	role, _ := roles[user]

	// Get the permissions for the role
	permissions, roleExists := acl[role]

	// If the role is not found in the ACL, deny access
	if !roleExists {
		return false
	}

	// Check if the permission exists in the role's permissions
	for _, p := range permissions {
		if p == permission {
			return true
		}
	}
	return false
}

func main() {
	// Get input from the command line
	var params struct {
		User       string `json:"user"`
		Permission string `json:"permission"`
	}

	if len(os.Args) < 2 {
		fmt.Println("Error: No input provided")
		os.Exit(1)
	}

	err := json.Unmarshal([]byte(os.Args[1]), &params)
	if err != nil {
		fmt.Println("Error parsing input:", err)
		os.Exit(1)
	}

	// Check access and display the result
	accessGranted := hasAccess(params.User, params.Permission)
	os.Stdout.WriteString(fmt.Sprintf("%t", accessGranted))
}
